File: Bob's Kingdom.py
# PYTHON IS TOO SLOW, CHECK C++ CODE
#
# https://dmoj.ca/problem/oly21practice9
# Binary search with greedy idea:
# - put everything with different >threshold in the other region
# - fill the bottom-right sub rectangles of every cell in the other region
#   - we can use a 2D PSA method
# - if the other region has (max-min) < threshold, it works
#
# Note: since there are 4 possible diagonal shapes (including symmetric ones)
#       we must try all of them

def works(grid, diff) -> bool:
    """Check if we can achieve a (max-min) of diff in both regions"""
    marked = [[False] * (C + 1) for _ in range(R + 1)]  # 1-indexed difference array
    for i in range(R):
        for j in range(C):
            # every cell with larger than diff difference must be in the other region
            if largest - grid[i][j] > diff:
                marked[i + 1][j + 1] = True

    # fill in the bottom-right sub-rectangles of marked cells to make diagonal
    for i in range(1, R + 1):
        for j in range(1, C + 1):
            marked[i][j] += marked[i - 1][j] + marked[i][j - 1] - marked[i - 1][j - 1]

    # special case: no cells marked means (max-min) threshold is already satisfied
    # and we can put any cell in the other region
    marked_cnt = sum(marked[i][j] != 0 for i in range(1, R + 1) for j in range(1, C + 1))
    if marked_cnt == 0:
        return True

    # check if the other region's difference is also <=diff
    small = min(grid[i][j] for i in range(R) for j in range(C) if marked[i + 1][j + 1])
    big = max(grid[i][j] for i in range(R) for j in range(C) if marked[i + 1][j + 1])
    return big - small <= diff


def check(diff) -> bool:
    """Check if any of the 4 orientations work"""
    return (works(grid, diff) or
            works(grid[::-1], diff) or
            works([r[::-1] for r in grid], diff) or
            works([r[::-1] for r in grid][::-1], diff))


R, C = map(int, input().split())
grid = [list(map(int, input().split())) for _ in range(R)]
largest = max(max(row) for row in grid)

low = 0
high = 10 ** 9
while low <= high:
    mid = (low + high) // 2
    if check(mid):
        high = mid - 1
    else:
        low = mid + 1

print(low)

File: Bob's Kingdom.cpp
// https://dmoj.ca/problem/oly21practice9
// requires stupid constant time optimizations, check python code for explanations

#pragma GCC optimize("O3,unroll-loops")
#pragma GCC target("avx2,bmi,bmi2,lzcnt,popcnt")
#include <bits/stdc++.h>
using namespace std;

const int MN = 2001;
int R, C;
int arr[MN][MN];
int largest;

bool works(int arr[MN][MN], int diff) {
    int marked[MN][MN] = {0};
    for (int i = 0; i < R; i++) {
        for (int j = 0; j < C; j++) {
            if (largest - arr[i][j] > diff) {
                marked[i + 1][j + 1] = 1;
            }
        }
    }

    for (int i = 1; i <= R; i++) {
        for (int j = 1; j <= C; j++) {
            marked[i][j] += marked[i - 1][j] + marked[i][j - 1] - marked[i - 1][j - 1];
        }
    }

    int marked_cnt = 0;
    for (int i = 1; i <= R; i++) {
        for (int j = 1; j <= C; j++) {
            if (marked[i][j] != 0) {
                marked_cnt++;
            }
        }
    }

    if (marked_cnt == 0) {
        return true;
    }

    int small = INT_MAX, big = INT_MIN;
    for (int i = 0; i < R; i++) {
        for (int j = 0; j < C; j++) {
            if (marked[i + 1][j + 1]) {
                small = min(small, arr[i][j]);
                big = max(big, arr[i][j]);
            }
        }
    }
    return big - small <= diff;
}

bool check(int diff) {
    int cop[MN][MN];
    memcpy(cop, arr, sizeof(arr));
    if (works(cop, diff)) return true;
    reverse(cop, cop + R);
    if (works(cop, diff)) return true;
    for (int i = 0; i < R; i++)
        reverse(cop[i], cop[i] + C);
    if (works(cop, diff)) return true;
    reverse(cop, cop + R);
    if (works(cop, diff)) return true;
    return false;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> R >> C;
    for (int i = 0; i < R; i++) {
        for (int j = 0; j < C; j++) {
            cin >> arr[i][j];
        }
    }

    largest = -1;
    for (int i = 0; i < R; i++) {
        for (int j = 0; j < C; j++) {
            largest = max(largest, arr[i][j]);
        }
    }

    int low = 0, high = 1e9;
    while (low <= high) {
        int mid = (low + high) / 2;
        if (check(mid)) {
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }

    cout << low << endl;

    return 0;
}
